core/backend/src/StashManager.ts

```diff
@@ -0,0 +1,458 @@
+    const stashDir = StashManager.getStashRootFolder(db, false);
+    if (!existsSync(stashDir)) {
+      return stashes;
+    }
+    readdirSync(stashDir).filter((file) => {
+      const filePath = path.join(stashDir, file);
+      if (existsSync(filePath) && statSync(filePath).isFile() && file.endsWith(".stash")) {
+        const id = file.slice(0, -path.extname(file).length)
+        const stash = this.getStash({ db, stash: id });
+        if (stash) {
+          stashes.push(stash);
+        }
+      }
+    });
+    stashes.sort((a, b) => Date.parse(b.timestamp) - Date.parse(a.timestamp));
+    return stashes;
+  }
+
+  /**
+   * Deletes the stash file associated with the specified stash ID or properties from the given {@link BriefcaseDb}.
+   *
+   * @param db - The {@link BriefcaseDb} instance from which the stash should be dropped.
+   * @param stashId - The unique identifier (GuidString) or properties (StashProps) of the stash to be deleted.
+   * @returns Returns `true` if the stash file was successfully deleted, otherwise returns `false`.
+   * @throws Does not throw; logs errors internally and returns `false` on failure.
+   */
+  public static dropStash(args: StashArgs): boolean {
+    try {
+      const stashFile = this.getStashFilePath(args);
+      unlinkSync(stashFile);
+      return true;
+    } catch (error: any) {
+      Logger.logError(loggerCategory, `Error dropping stash: ${error}`);
+    }
+    return false;
+  }
+
+  /**
+   * Removes all stashes associated with the specified {@link BriefcaseDb}.
+   *
+   * @param db - The {@link BriefcaseDb} instance from which all stashes will be removed.
+   */
+  public static dropAllStashes(db: BriefcaseDb): void {
+    this.getStashes(db).forEach((stash) => {
+      this.dropStash({ db, stash });
+    });
+  }
+
+  /**
+   * Queries the hub for the changeset information associated with the given stash.
+   *
+   * @param args - The arguments including the stash properties.
+   * @returns A promise resolving to the changeset ID and index.
+   */
+  private static async queryChangeset(args: StashArgs & { stash: StashProps }): Promise<ChangesetIdWithIndex> {
+    return IModelHost[_hubAccess].queryChangeset({
+      iModelId: args.stash.iModelId,
+      changeset: args.stash.parentChangeset,
+      accessToken: await IModelHost.getAccessToken()
+    });
+  }
+  /**
+   * Restores the specified stash to the given {@link BriefcaseDb}.
+   *
+   * This method will discard local changes, acquire required locks, pull and apply changesets if needed,
+   * and then restore the stash from the stash file.
+   *
+   * @param args - The arguments including the target database and stash properties.
+   * @throws IModelError if the restore operation fails.
+   */
+  public static async restore(args: StashArgs): Promise<void> {
+    const { db } = args;
+    Logger.logInfo(loggerCategory, `Restoring stash: ${this.getStashId(args)}`);
+
+    db.clearCaches();
+    db[_nativeDb].clearECDbCache();
```

Comment:
Should these two lines get moved down, so they don't happen at all in the event of some error being thrown?

Reply:
I should have removed it. I moved part of this as IModelDb.discardChanges() which already does all the cache clean up